File: content/LLD/Reference-Code/5. Strategy Pattern/discount_service.py
from discount_strategy import DiscountStrategy


class DiscountService:
    def __init__(self, discount_strategy: DiscountStrategy):
        self.__strategy = discount_strategy

    def set_strategy(self, new_discount_strategy: DiscountStrategy):
        self.__strategy = new_discount_strategy

    def process(self):
        self.__strategy.calculate_discount()

File: content/LLD/Reference-Code/5. Strategy Pattern/DiscountService.java
public class DiscountService {
    private DiscountStrategy strategy;

    public DiscountService(DiscountStrategy discountStrategy) {
        this.strategy = discountStrategy;
    }

    public void setStrategy(DiscountStrategy newDiscountStrategy) {
        this.strategy = newDiscountStrategy;
    }

    public void process() {
        this.strategy.calculateDiscount();
    }
}
